### PR TITLE
POP3, fix multi-line responses

### PR DESCRIPTION
**Pull request by icing**
Some POP3 commands are multi-line, e.g. have responses terminated by a last line with '.', but some are not. Define the known command properties and fix response handling.

Refs #14677, where the 'STAT' command is hanging as it waits for a multi-line response that never comes.

Add test case for STAT.

### CODE REVIEW

File: lib/pop3.c [hunk 1]
```diff
@@ -199,6 +203,47 @@ static void pop3_to_pop3s(struct connectdata *conn)
...
 #endif
 
+struct pop3_cmd {
+  const char *name;
+  unsigned short nlen;
+  BIT(multiline); /* response is multi-line with last '.' line */
+  BIT(multiline_with_args); /* is multi-line when command has args */
+};
+
+static struct pop3_cmd pop3cmds[] = {
```
Comment (bagder) [inline lib/pop3.c:213]:
```suggestion
static const struct pop3_cmd pop3cmds[] = {
```

Comment (icing) [reply lib/pop3.c]: Done.

**Comment by dfandrich** [comment]
Analysis of PR #14707 at [5bbeca0f](https://github.com/curl/curl/commit/5bbeca0fc0fa5d1ae9455b6604c91d09f68ba516):

[Test 310 failed](https://github.com/curl/curl/actions/runs/10595935131/job/29362791134#check-step-13), which has NOT been flaky recently, so **there could be a real issue in the PR**.

###### Generated by [Testclutch](https://testclutch.curl.se/)

**Comment by dfandrich** [comment]
NB, there are a few more POP3 commands out there.  RFC6856 defines UTF8, 
RFC2595 defines STLS, RFC5034 defines AUTH. There are also a number of
nonstandard extensions that appear to be available on servers in the wild,
defined in https://datatracker.ietf.org/doc/html/draft-hansen-pop3-xtndext-00,
namely "XTND XMIT" and "XTND XLST"

**Comment by bagder** [comment]
> there are a few more POP3 commands out there

That's exactly why `--request` does not typically offer this functionality. There is simply nothing that prevents someone from having a new command in their server and curl cannot magically be made to work with those commands. I think of it as a weakness in the protocol.

However, I think this fix from @icing is small and simple enough to be worth it as it extends the functionality a little bit at least, and it fixes this one use case that a user thought was an okay use case.

Perhaps also a little bit because POP3 is not really a protocol on the rise so maybe the need for doing many others is not that huge.

**Comment by icing** [comment]
We can think about what the default behaviour for unknown commands should be. I made it that single-line is the default. "XNTD" we can add as a multi-line. Or we change the defaults.

In either case, I think we should at least be able to handle the original RFC commands correctly.

**Comment by icing** [comment]
Latest push:

    add more known POP3 commands, change default

    Add more command definitions that we know of and change
    the default for unknown commands to multi-line, the behaviour
    curl had for all commands before this change.

I believe this PR is now backward compatible with the addition of handling known, single-line commands correctly.

**Review by bagder — APPROVED** [review]

**Comment by icing** [comment]
For usability it would be nice to dump the single-line command response as the body. Otherwise it would only be available via a `-D -` argument to curl, which then dumps all silly pop3 replies.

A matter of style points.

**Comment by bagder** [comment]
> For usability it would be nice to dump the single-line command response as the body

Maybe. That would be an even bigger "breakage" of established procedures though since it truly is a "header" that it returns.

**Pull request closed without merging**